File: kotlin/src/Day1.kt
import kotlin.math.abs

class Day1 : IDay {
    private fun getNumbers(line: String): Pair<Int, Int> {
        val numbers = line.split(Regex(" +")).map { it.toInt() }
        return numbers[0] to numbers[1]
    }

    override fun part1(input: String) : String {
        val items = input
            .lines()
            .map { line -> getNumbers(line) }

        val firstList = items.map { it.first }.sorted()
        val secondList = items.map { it.second }.sorted()

        return firstList.zip(secondList).sumOf { (first, second) ->
            abs(first - second)
        }.toString()
    }

    override fun part2(input: String) : String {
        val items = input
            .lines()
            .map { line -> getNumbers(line) }

        val firstList = items.map { it.first }
        val countMap = items.map { it.second }
            .fold(mutableMapOf<Int, Int>()) { countMap, value ->
                countMap[value] = (countMap[value] ?: 0) + 1
                countMap
            }

        return firstList.sumOf { value ->
            value * (countMap[value] ?: 0)
        }.toString()
    }
}

File: clojure/src/day1.clj
(ns day1
  (:require
   [clojure.string :refer [split-lines]]))

(defn get-distance [[first_list second_list]]
  (map #(abs (- %1 %2)) (sort first_list) (sort second_list)))

(defn create-lists [acc line]
  (->> (re-seq #"\d+" line)
       (map parse-long)
       (map conj acc)))

(defn part1 [input]
  (->> (split-lines input)
       (reduce create-lists [[], []])
       (get-distance)
       (reduce +)))

(defn list-to-count-map [list]
  (reduce
   (fn [counts, num]
     (assoc counts num (inc (get counts num 0))))
   {}
   list))

(defn get-similarity-score [[first_list second_list]]
  (let [counts (list-to-count-map second_list)]
    (map #(* % (get counts % 0)) first_list)))

(defn part2 [input]
  (->> (split-lines input)
       (reduce create-lists [[], []])
       (get-similarity-score)
       (reduce +)))